### PR TITLE
Add Sandbox test case generator example - 2nd pass

### PR DESCRIPTION
**Pull request by zhang-lucy**
⚡ Changes since last PR (https://github.com/modal-labs/modal-examples/pull/1280)
- Prompt Engineering - Tried a variety of different techniques - surprisingly,  the repetitive hints seem to be required for keeping the code working. I'll keep iterating a bit more but would prefer to keep this relatively error-free
- Sandbox Bells-and-Whistles: Added 2 tunnels - one for showing the code diff, one for [Allure](https://allurereport.org/) - cute testing report framework. See screenshots!
- SGLang: Made it the client-server model as [recommended](https://modal-com.slack.com/archives/C08G83NB9B3/p1753288465186799).
- Various cleanups + responded to comments 

‼️ Outstanding Issues
- Error is thrown with the final `asyncio.gather`, as the `.wait()` function throws an error on timeout. Will be fixed in subsequent PR
- Volumes should be ephemeral

💡 Outstanding Q's
- Overall Runtime: ~3:20. Can we do better? The server startup time is the most obvious place to cut down on. Taking any recs!
    - Image build ~60 sec
    - Server startup time ~75 sec
    - Inference ~45 sec
    - Sandboxes ~20 sec
- Deepseek Coder-6.7B-Instruct? Originally I picked this model since it was small and seemed to do pretty well on [the leaderboard](https://huggingface.co/spaces/bigcode/bigcode-models-leaderboard). Could try a bigger model if we're doing client-server.
- How much retry + additional validation logic should we add? The current generate function will error if the length of the output exceeds the set limit - this is flaky as it's really prompt driven
- Edit: Just realized I didn't add the ephemeral volumes piece. Will do in a subsequent push.

❓ General high level question is just how much optimization to do + what's missing here? The focus of this example is _Sandboxes_; we have plenty of inference examples elsewhere.

image
image
image
image

## Type of Change

- [x] New example for the GitHub repo

## Monitoring Checklist

  - [x] Example is configured for testing in the synthetic monitoring system, or `lambda-test: false` is provided in the example frontmatter and I have gotten approval from a maintainer
    - [x] Example is tested by executing with `modal run`, or an alternative `cmd` is provided in the example frontmatter (e.g. `cmd: ["modal", "serve"]`)
    - [x] Example is tested by running the `cmd` with no arguments, or the `args` are provided in the example frontmatter (e.g. `args: ["--prompt", "Formula for room temperature superconductor:"]`
    - [x] Example does _not_ require third-party dependencies besides `fastapi` to be installed locally (e.g. does not import `requests` or `torch` in the global scope or other code executed locally)

### Build Stability
  - [x] Example pins all dependencies in container images
    - [x] Example pins container images to a stable tag like `v1`, not a dynamic tag like `latest`
    - [x] Example specifies a `python_version` for the base image, if it is used 
    - [x] Example pins all dependencies to at least [SemVer](https://semver.org/) minor version, `~=x.y.z` or `==x.y`, or we expect this example to work across major versions of the dependency and are committed to maintenance across those versions
      - [x] Example dependencies with `version < 1` are pinned to patch version, `==0.y.z`

## Outside Contributors

You're great! Thanks for your contribution.

### CODE REVIEW

File: 13_sandboxes/test_case_generator.py [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+        wait_for_port(self.serve_process, 8000)
+
+        print("SGLang server is ready!")
+
+    @modal.web_server(port=8000, startup_timeout=240)
+    def serve(self):
+        return
+
+    @modal.method()
+    def healthcheck(self):
```
Comment (zhang-lucy) [inline 13_sandboxes/test_case_generator.py]: this isn't called but feels right to have

File: 13_sandboxes/test_case_generator.py [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+            response_format={
+                "type": "json_schema",
+                "json_schema": {
+                    "name": "test_file",
+                    "schema": json_schema,
+                },
+            },
+        )
+        output = response.choices[0].message.content
+        try:
```
Comment (zhang-lucy) [inline 13_sandboxes/test_case_generator.py]: open to add'l suggestions on error handling and/or retries

File: 13_sandboxes/test_case_generator.py [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+    sb = modal.Sandbox.create(
+        "sh",
+        "-c",
+        cmd,
+        app=app,
+        image=image,
+        volumes={
+            "/data": files_volume,
+        },
+        timeout=300,  # TODO: should we keep this as the default?
```
Comment (zhang-lucy) [inline 13_sandboxes/test_case_generator.py]: will remove

File: 13_sandboxes/test_case_generator.py [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+        tunnel1 = sb.tunnels()[8000]
+        tunnel2 = sb.tunnels()[8001]
+        print(f"Sandbox created and run for generated test file: {filename}")
+        print(f"✨ View diff: {tunnel2.url}")
+        print(f"✨ View test results: {tunnel1.url}\n")
+
+    return file_to_sandbox.values()
+
+
+def poll_sandboxes(sandboxes: list[modal.Sandbox]):
```
Comment (zhang-lucy) [inline 13_sandboxes/test_case_generator.py]: hacky; keeps the app alive so it doesn't kill all the sandboxes

File: 13_sandboxes/test_case_generator.py [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+# # Addenda
+# The below functions are utility functions.
+def create_sandboxes(filenames: list[str], gh_owner: str, gh_repo_name: str):
+    file_to_sandbox: dict[str, modal.Sandbox] = {}
+    for filename in filenames:
+        print(f"Running sandbox for {filename}")
+        image = get_sandbox_image(gh_owner, gh_repo_name)
+        sb = run_sandbox(image, filename)
+        file_to_sandbox[filename] = sb
+    time.sleep(20)
```
Comment (zhang-lucy) [inline 13_sandboxes/test_case_generator.py]: hack to make sure that the tunnels show up at the very end, after the noisy sandbox logs. don't think there's a sane way to silence those logs, but i feel like it's ok? kinda cool to see pytest running in the background in the sandboxes.

File: 13_sandboxes/test_case_generator.py [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+    3. Based on that understanding, **add** new test cases to the test file to increase coverage—especially edge cases, boundary conditions, and untested branches.
+    4. Use `pytest` idioms, but do **not** add or change import statements—**use only what is already imported**.
+    5. Do **not** explain your reasoning—just return the final modified test file.
+
+    ### Requirements:
+    - Your output must be a valid, complete Python file with the added test cases.
+    - Do not modify existing test logic unless necessary to support your new test cases.
+    - Do not import any additional modules.
+    - Limit each line to a maximum of 100 characters to avoid output truncation or formatting errors.
+    - Limit your output to around 25 lines. Make sure to complete any functions or blocks you start.
```
Comment (zhang-lucy) [inline 13_sandboxes/test_case_generator.py:325]: this particular line needs to be present in both the system and user prompt to minimize flakiness. will experiment with what other lines i can remove/de-dupe.

File: 13_sandboxes/test_case_generator.py [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+model_volume = modal.Volume.from_name("deepseek-model-volume", create_if_missing=True)
+files_volume = modal.Volume.from_name("files-volume", create_if_missing=True)
+
+MODEL_NAME = "deepseek-ai/deepseek-coder-6.7b-instruct"
+MODEL_REVISION = "e5d64addd26a6a1db0f9b863abf6ee3141936807"
+
+
+model_image = (
+    modal.Image.from_registry("lmsysorg/sglang:v0.4.9.post3-cu126", add_python="3.12")
+    .uv_pip_install(
```
Comment (pawalt) [inline 13_sandboxes/test_case_generator.py:22]: Why do all these deps need to be installed if we're using the sglang image?

File: 13_sandboxes/test_case_generator.py (hunk 1 above)
Comment (pawalt) [reply 13_sandboxes/test_case_generator.py]: I don't think we should have it if it's not used. It's the job of some calling application to health check anyways.

File: 13_sandboxes/test_case_generator.py [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+)
+@modal.concurrent(max_inputs=3)  # Each container runs up to 3 requests at once.
+class TestCaseServer:
+    @modal.enter()
+    def download_model(self):
+        from huggingface_hub import snapshot_download
+
+        snapshot_download(
+            MODEL_NAME,
+            local_dir=f"/cache/{MODEL_NAME}",
```
Comment (pawalt) [inline 13_sandboxes/test_case_generator.py]: Does `local_dir` work like `cache_dir`?

**Review by pawalt — COMMENTED** [review]
I think this needs to be written in the literate programming style - this is just code right now.

**Comment by zhang-lucy** [comment]
> I think this needs to be written in the literate programming style - this is just code right now.

@pawalt this PR will be just code - will add in the narrative text in a follow-up PR

File: 13_sandboxes/test_case_generator.py (hunk 7 above)
Comment (zhang-lucy) [reply 13_sandboxes/test_case_generator.py:22]: removed some, kept the ones for hf transfer

File: 13_sandboxes/test_case_generator.py (hunk 8 above)
Comment (zhang-lucy) [reply 13_sandboxes/test_case_generator.py]: yup! https://huggingface.co/docs/huggingface_hub/en/guides/download#download-files-to-a-local-folder

**Comment by charlesfrye** [comment]
🚀 The docs preview is ready! Check it out here:     https://modal-labs-examples--frontend-preview-6dc4477.modal.run

**Comment by charlesfrye** [comment]
> I think this needs to be written in the literate programming style - this is just code right now.

Quick note here -- I'd actually prefer folks write new examples with only code + comments the first time around, then we can do the full literate programming/documentation site treatment afterwards before/when we promote the example to become discoverable.

File: 13_sandboxes/test_case_generator.py (hunk 8 above)
Comment (pawalt) [reply 13_sandboxes/test_case_generator.py]: Can you add a comment to this effect?

File: 13_sandboxes/test_case_generator.py (hunk 2 above)
Comment (pawalt) [reply 13_sandboxes/test_case_generator.py]: Can you just raise here? Or is it too noisy? If it's too noisy, I think you should add a 5 second, 10 retry system. We shouldn't swallow errors.

File: 13_sandboxes/test_case_generator.py (hunk 5 above)
Comment (pawalt) [reply 13_sandboxes/test_case_generator.py]: I also don't have a good suggestion here. Can you add a `HACK` comment?

File: 13_sandboxes/test_case_generator.py (hunk 4 above)
Comment (pawalt) [reply 13_sandboxes/test_case_generator.py]: Can you just `asynctio.join` on `sb.wait.aio`?

File: 13_sandboxes/test_case_generator.py (hunk 2 above)
Comment (zhang-lucy) [reply 13_sandboxes/test_case_generator.py]: Changed it to just raise here - I think structured inputs should be fine

File: 13_sandboxes/test_case_generator.py (hunk 5 above)
Comment (zhang-lucy) [reply 13_sandboxes/test_case_generator.py]: Added

**Comment by charlesfrye** [comment]
🚀 The docs preview is ready! Check it out here:     https://modal-labs-examples--frontend-preview-496205a.modal.run

**Comment by charlesfrye** [comment]
🚀 The docs preview is ready! Check it out here:     https://modal-labs-examples--frontend-preview-1ac2004.modal.run

**Comment by charlesfrye** [comment]
🚀 The docs preview is ready! Check it out here:     https://modal-labs-examples--frontend-preview-b63d469.modal.run

**Comment by zhang-lucy** [comment]
This example passes all checks except for the final line:

```python
await asyncio.gather(*[sb.wait.aio(raise_on_termination=False) for sb in sandboxes])
```

because `.wait()` default throws an error (`raise_on_termination=False` doesn't prevent the error being raised on timeout). Merging because there's been enough iteration on this example; will fix this in a follow-up PR. Also updated description.
